File: apps/backend/src/routes/ai.js
import { Hono } from 'hono';
import { aiClient } from '../../ai/AIClient.js';
import { EntitlementService } from '../services/EntitlementService.js';
import { db } from '../db/index.js';
import { secretService } from '../services/SecretService.js';
import { spreadsheetToolService } from '../services/SpreadsheetToolService.js';
import { authMiddleware } from '../middleware/auth.js';

const aiRoutes = new Hono();
aiRoutes.use('*', authMiddleware);
const entitlementService = new EntitlementService(db);

/**
 * GET /config
 * Get current AI configuration (User Preference)
 */
aiRoutes.get('/config', async (c) => {
    try {
        const userId = c.get('userId') || c.req.header('x-user-id');
        if (!userId) return c.json({ error: 'Unauthorized' }, 401);

        const vaultKey = `secret/pegasus/users/${userId}/ai/config`;
        const configStr = await secretService.resolveSecret(`vault://${vaultKey}`);

        return c.json(configStr ? JSON.parse(configStr) : { provider: 'default' });
    } catch (error) {
        console.error('[AI Config] Error getting config:', error);
        return c.json({ error: 'Failed to fetch AI config' }, 500);
    }
});

/**
 * POST /config
 * Set AI Provider preference (Gated feature)
 */
aiRoutes.post('/config', async (c) => {
    try {
        const userId = c.get('userId') || c.req.header('x-user-id');
        if (!userId) return c.json({ error: 'Unauthorized' }, 401);

        const { provider, modelId } = await c.req.json();

        // 1. Enterprise Gate Check
        if (provider !== 'default') {
            const hasAccess = await entitlementService.hasFeature(userId, 'byom_models');
            if (!hasAccess) {
                return c.json({
                    error: 'Custom AI Models are available on Teams & Enterprise plans only.',
                    upgradeRequired: true
                }, 403);
            }
        }

        // 2. Save Preference
        const vaultKey = `secret/pegasus/users/${userId}/ai/config`;
        await secretService.storeSecret(vaultKey, JSON.stringify({
            provider,
            modelId, // e.g. 'aws:anthropic.claude-3...'
            updated_at: Date.now()
        }));

        return c.json({ success: true, message: 'AI preference updated' });
    } catch (error) {
        console.error('[AI Config] Error setting config:', error);
        return c.json({ error: 'Failed to update AI config' }, 500);
    }
});

/**
 * GET /models
 * List available models based on user tier and connections
 */
aiRoutes.get('/models', async (c) => {
    try {
        const userId = c.get('userId') || c.req.header('x-user-id');
        if (!userId) return c.json({ error: 'Unauthorized' }, 401);

        // Fetch models from AIClient (which now checks connected accounts)
        const models = await aiClient.listModels(userId);

        // Filter based on entitlement
        const tier = await entitlementService.getTier(userId);
        const hasByomAccess = await entitlementService.hasFeature(userId, 'byom_models');

        const allowedModels = models.filter(m => {
            if (m.id === 'gemini-2.5-flash' || m.id === 'gemini-2.5-pro') {
                return tier === 'free';
            }
            if (m.id === 'gemini-3-flash-preview' || m.id === 'gemini-3-pro-preview') {
                return tier !== 'free';
            }
            return true;
        });

        const visibleModels = allowedModels.map(m => {
            const isByom = m.id.startsWith('aws:') || m.id.startsWith('azure:') || m.id.startsWith('gcp:');
            return {
                ...m,
                locked: isByom && !hasByomAccess
            };
        });

        return c.json({ models: visibleModels });
    } catch (error) {
        console.error('[AI Models] Error listing models:', error);
        return c.json({ error: 'Failed to list models', models: [] }, 500);
    }
});

/**
 * POST /query
 * Execute a structured AI command for Data Studio
 */
aiRoutes.post('/query', async (c) => {
    try {
        const userId = c.get('userId') || c.req.header('x-user-id') || 'dev_user';
        const { query } = await c.req.json();

        if (!query) return c.json({ error: 'Query is required' }, 400);

        // Hardcode to gemini-3-flash-preview for efficiency as requested
        const targetModel = 'gemini-3-flash-preview';

        console.log(`[AI Query] Executing for ${userId}: ${query.substring(0, 50)}...`);

        const response = await aiClient.generateContent([
            { role: 'user', content: query }
        ], {
            model: targetModel,
            userId,
            json: true // Ensure we get JSON back if possible
        });

        return c.json({
            result: response.text || response,
            usage: response.usage
        });
    } catch (error) {
        console.error('[AI Query] Error:', error);
        return c.json({ error: 'AI request failed', details: error.message }, 500);
    }
});

export default aiRoutes;

File: apps/python-intelligence/app/routes/ai.py
from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException

from app.models import (
    AIConfigRequest,
    AIModelInfo,
    AIQueryRequest,
    AIQueryResponse,
    GenerateRequest,
    GenerateResponse,
)
from app.routes.auth import get_current_user
from app.services.provider import AIProviderClient

router = APIRouter(prefix="/ai", tags=["ai"])
client = AIProviderClient()


@router.get("/config")
async def get_ai_config(user: dict = Depends(get_current_user)):
    # In production, fetch from Vault/Secrets
    return {"provider": "default"}


@router.post("/config")
async def set_ai_config(req: AIConfigRequest, user: dict = Depends(get_current_user)):
    # Check entitlements for custom providers
    if req.provider != "default":
        tier = user.get("subscription_tier", "free")
        if tier not in ("pro_plus", "enterprise", "teams"):
            raise HTTPException(
                status_code=403,
                detail="Custom AI Models available on Teams & Enterprise plans only",
            )
    # Save to Vault in production
    return {"success": True, "message": "AI preference updated"}


@router.get("/models", response_model=dict)
async def list_models(user: dict = Depends(get_current_user)):
    """List available models based on user tier."""
    tier = user.get("subscription_tier", "free")
    models = await client.list_models()

    # Filter by tier
    free_only = {"gemini-2.5-flash", "gemini-2.5-pro"}
    pro_models = {"gemini-3-flash-preview", "gemini-3-pro-preview"}
    byom_prefixes = ("aws:", "azure:", "gcp:")

    has_byom = tier in ("pro_plus", "enterprise", "teams")
    restricted: list[AIModelInfo] = []
    for m in models:
        if tier == "free" and m.id not in free_only:
            continue
        if tier != "free" and m.id in free_only:
            continue
        m.locked = m.id.startswith(byom_prefixes) and not has_byom
        restricted.append(m)

    return {"models": [m.model_dump(by_alias=True) for m in restricted]}


@router.post("/query", response_model=AIQueryResponse)
async def ai_query(req: AIQueryRequest, user: dict = Depends(get_current_user)):
    """Execute a structured AI command."""
    if not req.query:
        raise HTTPException(status_code=400, detail="Query is required")

    result = await client.generate_text(
        prompt=req.query,
        model="gemini-3-flash-preview",
        user_id=user["id"],
    )
    return AIQueryResponse(result=result.text, usage=result.usage)


@router.post("/generate", response_model=GenerateResponse)
async def generate(req: GenerateRequest, user: dict = Depends(get_current_user)):
    """Generate content from messages."""
    result = await client.generate_content(
        messages=[m.model_dump() for m in req.messages],
        model=req.model,
        temperature=req.temperature,
        max_tokens=req.max_tokens,
        json_mode=req.json,
        user_id=req.user_id or user["id"],
    )
    return GenerateResponse(text=result.text, usage=result.usage)


@router.post("/dashboard-query", response_model=AIQueryResponse)
async def dashboard_query(req: AIQueryRequest, user: dict = Depends(get_current_user)):
    """AI query triggered from dashboard chat."""
    result = await client.generate_text(
        prompt=req.query,
        model="gemini-3-flash-preview",
        user_id=user["id"],
    )
    return AIQueryResponse(result=result.text, usage=result.usage)
